cmd/get-kustomize-build-base/main.go

```diff
@@ -0,0 +1,141 @@
+package main
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"strings"
+	"syscall"
+
+	"github.com/urfave/cli/v2"
+)
+
+// see https://git-scm.com/docs/git-diff#_raw_output_format
+const (
+	changeStatusModification = "M"
+	changeStatusChangingType = "T"
+	changeStatusAddition     = "A"
+	changeStatusDeletion     = "D"
+	changeStatusCopy         = "C"
+	changeStatusRenaming     = "R"
+	// not relevant: U and X
+
+	gitDiffStatusColumn = 4
+)
+
+func main() { //go-cov:skip
+	if err := runApp(context.Background(), os.Stdout, os.Stdin, os.Args); err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(1)
+	}
+}
+
+func runApp(ctx context.Context, out io.Writer, in io.Reader, args []string) error {
+	ctx, cancel := signal.NotifyContext(ctx, syscall.SIGINT)
+	defer cancel()
+
+	app := buildApp(out, in)
+
+	return app.RunContext(ctx, args)
+}
+
+func buildApp(out io.Writer, in io.Reader) cli.App {
+	return cli.App{
+		Name:  "get-kustomize-build-base",
+		Usage: "Print the list of build bases for kustomize given the output of `git diff --raw` provided via stdin",
+		Flags: []cli.Flag{
+			&cli.StringFlag{
+				Name:     "build-root",
+				Required: true,
+			},
+		},
+		Action: func(c *cli.Context) error {
+			return run(c.Context, out, in, c.String("build-root"))
+		},
+	}
+}
+
+func run(_ context.Context, out io.Writer, in io.Reader, buildRoot string) error {
+	scanner := bufio.NewScanner(in)
+	for scanner.Scan() {
+		line := scanner.Text()
+		parts := strings.Split(line, " ")
+		// files are listed directly after the status
+		statusAndFiles := strings.Split(parts[gitDiffStatusColumn], "\t")
+		status := statusAndFiles[0]
+		files := statusAndFiles[1:]
+		kustomizeBase, err := getBaseForChange(buildRoot, status, files)
+		if err != nil { //go-cov:skip
+			return err
+		}
+		if kustomizeBase != "" {
+			fmt.Fprintln(out, kustomizeBase)
+		}
+	}
+	if err := scanner.Err(); err != nil { //go-cov:skip
+		return fmt.Errorf("scanning input: %w", err)
+	}
+
+	return nil
+}
+
+func getBaseForChange(buildRoot string, status string, files []string) (string, error) {
+	switch status {
+	case changeStatusModification, changeStatusChangingType:
+		return findKustomizationRoot(buildRoot, files[0])
+	case changeStatusAddition, changeStatusDeletion:
+		return findKustomizationRootForNew(buildRoot, files[0])
```

Comment:
So, if we are adding/ deleting/ copying/ renaming Kustomization, then we look for the `parent` Kustomization file (I suppose, to see if it was included there).
But, if we are modifying the Kustomization, then we don't look for the parent- we take just that. 

Why shouldn't we always look for the `parent`?

Reply:
Also, currently, if somebody adds parent level kustomization, this tool won't find any changes.

Reply:
Maybe to solve the problem for all the cases, we could just revert the algorthim for `findKustomizationRoot` - instead of going down the path, we could go up the path from repository level. 
This would mean, we don't need whole of that algorithm to parse the git change type. 
WDYT?

Reply:
> Maybe to solve the problem for all the cases, we could just revert the algorthim for findKustomizationRoot - instead of going down the path, we could go up the path from repository level.

I'm not sure I follow, could you expand on what you mean?

Reply:
If the change now is at `a/b/c/d/file.txt`, we look for kustomization.yaml in
`a/b/c/d`
`a/b/c`
`a/b`
`a`

What if we would look for it in
`a`
`a/b`
`a/b/c`
`a/b/c/d`
?

We would always find the first parent then, and we capture all the changes.

Reply:
I assume, if this would be correct, then we would just need to modify the `findKustomizationRoot` function in `kustomize-build-dirs` to resolve the issue

Reply:
This demands assumption, that each kustomization.yaml above in hierarchy is parent- but we operate with this assumption now

Reply:
> If the change now is at `a/b/c/d/file.txt`, we look for kustomization.yaml in `a/b/c/d` `a/b/c` `a/b` `a`
> 
> What if we would look for it in `a` `a/b` `a/b/c` `a/b/c/d` ?
> 
> We would always find the first parent then, and we capture all the changes.

:thinking: in this case I think we'll be exclusively building entire namespaces on each change/run. I expect this would slow things down since we'd need to do a bunch of cloning to fetch remove kustomizations that we don't actually 'need' (i.e. from building all the non-changed services in that namespace). Here's a run rebuilding all of (I think one of the largest?) namespaces https://github.com/utilitywarehouse/kubernetes-manifests/actions/runs/10616576962/job/29427336227 it took ~3 minutes. For other namespaces though this should be much quicker. It's probably worth experimenting with